### PR TITLE
SI-8173 add support for patterns like init :+ last to quasiquotes

### PR DESCRIPTION
**Pull request by densh**
Adds support for patterns like:

``` scala
  val q"{ ..$init; $last }" = q"{ a; b; c }"
  // init == List(q"a", q"b")
  // last == q"c"
```

Which under the hood get compiled as `:+` patterns:

``` scala
  SyntacticBlock(init :+ last)
```

### CODE REVIEW

**Comment by densh** [comment]
review by @retronym @xeno-by

File: src/compiler/scala/tools/reflect/quasiquotes/Reifiers.scala [hunk 1]
```diff
@@ -305,7 +305,7 @@ trait Reifiers { self: Quasiquotes =>
      *    > reifyMultiCardinalityList(lst) { ... } { ... }
      *    q"List($foo, $bar) ++ ${holeMap(qq$f3948f9s$1).tree}"
      */
-    def reifyMultiCardinalityList[T](xs: List[T])(fill: PartialFunction[T, Tree])(fallback: T => Tree): Tree
+    def reifyMultiCardinalityList(xs: List[Any])(fill: PartialFunction[Any, Tree])(fallback: Any => Tree): Tree
```
Comment (densh) [inline src/compiler/scala/tools/reflect/quasiquotes/Reifiers.scala:308]: It turned out that all usages of this function always had `T = Any` so there was little utility of leaving it generic.

File: test/files/scalacheck/quasiquotes/QuasiquoteProperties.scala [hunk 2]
```diff
@@ -116,10 +116,5 @@ trait Helpers {
     }
   }
 
-  def annot(name: String): Tree = annot(TypeName(name), Nil)
-  def annot(name: TypeName): Tree = annot(name, Nil)
-  def annot(name: String, args: List[Tree]): Tree = annot(TypeName(name), args)
-  def annot(name: TypeName, args: List[Tree]): Tree = q"new $name(..$args)"
-
```
Comment (densh) [inline test/files/scalacheck/quasiquotes/QuasiquoteProperties.scala:123]: Removing as utility over q"new ..." is minimal.

File: src/compiler/scala/tools/reflect/quasiquotes/Reifiers.scala [hunk 3]
```diff
@@ -403,14 +403,26 @@ trait Reifiers { self: Quasiquotes =>
...
+      val grouped = group(xs) { (a, b) => !fill.isDefinedAt(a) && !fill.isDefinedAt(b) }
+      def appended(lst: List[Any], init: Tree)  = lst.foldLeft(init)  { (l, r) => append(l, fallback(r)) }
+      def prepended(lst: List[Any], init: Tree) = lst.foldRight(init) { (l, r) => cons(fallback(l), r)   }
+      grouped match {
+        case init :: List(hole) :: last :: Nil if fill.isDefinedAt(hole) => appended(last, prepended(init, fill(hole)))
+        case init :: List(hole) :: Nil         if fill.isDefinedAt(hole) => prepended(init, fill(hole))
+        case         List(hole) :: last :: Nil if fill.isDefinedAt(hole) => appended(last, fill(hole))
+        case         List(hole) :: Nil         if fill.isDefinedAt(hole) => fill(hole)
+        case _                                                           => prepended(xs, collectionNil)
+      }
```
Comment (densh) [inline src/compiler/scala/tools/reflect/quasiquotes/Reifiers.scala:425]: These patterns would have been a lot nicer if `PartialFunction` had an `def unapply(arg: A): Option[B] = lift(arg)` method:

``` scala
grouped match {
  case init :: List(Fill(hole)) :: last :: Nil => appended(last, prepended(init, hole))
  case init :: List(Fill(hole)) :: Nil         => prepended(init, hole)
  case         List(Fill(hole)) :: last :: Nil => appended(last, hole)
  case         List(Fill(hole)) :: Nil         => hole
  case _                                       => prepended(xs, collectionNil)
}
```

**Comment by densh** [comment]
ping @xeno-by

**Comment by xeno-by** [comment]
LGTM

**Pull request merged**